### ISSUE TITLE
ActivityPub Create Activity with object reference

### ISSUE THREAD

**Issue by bmottershead** [mastodon/mastodon#19102]
### Steps to reproduce the problem

If my AP server (under development) sends a Create Activity to a mastodon inbox, which as the Note embedded in the
activity, as below, then it seems to work, in the sense that the note appears in the addressee's timeline on mastodon.social, almost immediately.  However, if the created Note is not embedded, but referenced, then it does not seem to work.  That is, nothing is visible on mastodon.social. The POST to the mastodon.social actor's inbox returns a 202 in both cases.

Create activity with "embedded" object (simplified)

```json
{
"@context":"https://www.w3.org/ns/activitystreams",
"id":"https://example.org/activity/1", 
"type":"Create", 
"actor":"https://example.org/actor/1",
"to":"https://mastodon.social/users/someuser",
"object":{"id":"https://example.org/notes/1","type":"Note","content":"Hello, world"}
}
```

Create activity with "referenced" object (simplified)

```json
{
"@context":"https://www.w3.org/ns/activitystreams",
"id":"https://example.org/activity/1", 
"type":"Create", 
"actor":"https://example.org/actor/1",
"to":"https://mastodon.social/users/someuser",
"object":"https://example.org/notes/1"
}
```

### Expected behaviour

From my reading of the ActivityPub spec, I would have expected that both the "referenced" and "embedded" cases would have the same effect, with there being a GET by mastodon.social in the "referenced" case to retrieve the created note.

### Actual behaviour

The actual result is that only the "embedded" case seems to work, and there is no evidence in the logs that in the "referenced" case, mastodon.social attempted to dereference the URI of the created note.

### Specifications

Deployed version of mastodon.social as of Sep 1, 2022.

**Comment by bmottershead** [comment]
Interestingly, if I create the Note on my server, and then send an "Announce" to mastodon.social referencing the Note object, rather than a "Create", the Note appears in the addressee's timeline on mastodon.social, as expected.   

```json
{
"@context":"https://www.w3.org/ns/activitystreams",
"id":"https://example.org/activity/1", 
"type":"Announce", 
"actor":"https://example.org/actor/1",
"to":"https://mastodon.social/users/someuser",
"object":"https://example.org/notes/1"
}
```
I would expect that "Create"  and "Announce" in an inbox would both work essentially the same way on a federated Mastodon server, other than distinguishing the latter as a "boost".  But there isn't really much difference between them for a downstream server.   

The differences are only relevant to the original server, which of course must actually create the object and assign it an id in the "Create" case, but only pass along a reference to it in the "Announce" case.  But why do downstream servers care about the difference between Create and Announce?

**Comment by trwnh** [comment]
it looks like an oversight in app/lib/activitypub/activity/create.rb compared to app/lib/activitypub/activity/announce.rb -- the Announce logic uses `value_or_id`which can handle a string id for `object`, but the Create logic will first reject the payload if `object` is an unsupported type. unfortunately it looks like one of the unsupported object types is a string value. so this conditional check should probably be refactored.

**Comment by bmottershead** [comment]
Thank you for confirming that.   It would be interesting to know whether any other activities are handled differently depending on whether the "object" is referenced or nested/embedded.    Also, with nested objects, one sometimes sees "abbreviated" nested objects in online ActivityPub.   For example, a "abbreviated" nesting with Create might look like:

```json
{
"@context":"https://www.w3.org/ns/activitystreams",
"id":"https://example.org/activity/1", 
"type":"Create", 
"actor":"https://example.org/actor/1",
"to":"https://mastodon.social/users/someuser",
"object":{"id":"https://example.org/notes/1","type":"Note"}
}
```
I haven't tried it, but in theory this should also work, resulting in the downstream AP server retrieving the note in order to obtain the content (and other details)

**Comment by trwnh** [comment]
it would probably depend on the json-ld processing done by each project, and whether any fetch-based validation is done or if the embedded object is simply accepted as-is.

**Comment by bmottershead** [comment]
How about Mastodon, though?

**Comment by Gargron** [comment]
Mastodon wouldn't know that such an object is abbreviated and needs to be fetched. It would try to process it as-is and then invalidate it because of missing properties.

**Comment by bmottershead** [comment]
In the Activity Vocabulary recommedation, the "object" property is not marked as "Functional".   Functional properties are those that can take only one value; so not being marked as "Functional" means that "object" can have multiple values.  That means that activities like this are well formed:
```json
{
"@context":"https://www.w3.org/ns/activitystreams",
"id":"https://example.org/activity/1", 
"type":"Create", 
"actor":"https://example.org/actor/1",
"to":"https://mastodon.social/users/someuser",
"object":[
  "https://example.org/notes/1", 
  {"id":"https://example.org/notes/2","type":"Note","content":"Hello"},
  ...
]
}
```
This would be an actor creating several Notes in one go.  I am probably beating a dead horse here, but I am guessing that this isn't allowed by Mastodon, either.

**Comment by trwnh** [comment]
using one activity on multiple objects would be a separate issue, though -- i'd recommend testing it to see and then filing a bug if indeed it is not supported

**Comment by bmottershead** [comment]
Will do.  You are of course right that it is a separate issue.   As is the "abbreviated" case.   Thanks.

**Comment by nightpool** [comment]
I'm not even sure what it would mean semantically to create two notes in the same activity. Would they be linked posts in some way? Thinking about how to write a renderer for that sort of activity seems difficult.

**Comment by bmottershead** [comment]
Given infinite developer creativity, someone will come up with a use-case for it, if it is legal.

The bright side is that if Mastodon doesn't handle some of these cases on the receiving side, it won't send them either.  So your server can't generate it, but you also don't have to deal with it, either.

**Comment by trwnh** [comment]
> I'm not even sure what it would mean semantically to create two notes in the same activity. Would they be linked posts in some way? Thinking about how to write a renderer for that sort of activity seems difficult.

could just be used to create two statuses. maybe S2S batching? "tweet storm" functionality like on twitter where you can compose an entire thread at the same time?

**Comment by michaelcpuckett** [comment]
Hi - just ran into this same issue. I'm embedding the object now, but I would prefer to reference it.

**Comment by steve-bate** [comment]
I also ran into this issue. The `Create.perform` dereferences the `@json["object"]` into the `@object` field. However, later in `process_status_params`, the `StatusParser` is provided with the original `@json` data that still has the object IRI. This causes data like `text` to be blank and causes a subsequent failure in the ActiveRecord validation (basically a silent failure, unless debug logging is enabled).

It looks like it could be fixed by either overwritting `@json["object"]` in `perform` with the dereferenced object or by changing the `StatusParser` to use both `@json` and `@object`.

**Comment by steve-bate** [comment]
> In the Activity Vocabulary recommedation, the "object" property is not marked as "Functional".

That is true and I had the same thoughts. However, after discussing it with some of the ActivityPub spec authors, they claim that the AP spec effectively constrains the AS2 vocabulary in a way that makes `object` functional. They reference language in the AP specification like 'an object' and 'the object'.

**Comment by melvincarvalho** [comment]
> That is true and I had the same thoughts. However, after discussing it with some of the ActivityPub spec authors, they claim that the AP spec effectively constrains the AS2 vocabulary in a way that makes object functional. They reference language in the AP specification like 'an object' and 'the object'.

Thank you for sharing this information, it's quite interesting. However, I respectfully beg to differ with the interpretation suggested by the ActivityPub spec authors. The terms 'an object' and 'the object' in the AP specification don't necessarily imply functional constraints on the AS2 vocabulary.

It's worth noting that the use of both singular and plural forms is quite common, for instance "friend" vs "friends" or "relay" vs "relays". Moreover, a property term often corresponds to a Class with an uppercase letter, such as the Class 'Alias' corresponding to the property 'alias'. Although these terms are singular in English, they don't impose a limit on the quantity unless specifically marked as functional.

**Comment by melvincarvalho** [comment]
> "object":{"id":"https://example.org/notes/1","type":"Note","content":"Hello, world"}

Your point on treating URIs as definitive references is intriguing. On the web, URIs are indeed seen as pointers with their current contents regarded as authoritative. When posting a URI to an endpoint, it might carry additional properties for practicality, yet the original URI is deemed as the ultimate source of truth. The presence of a signature lends more credibility to the posted data.

However, implementation practices may differ based on practical considerations, such as the cost of lookups and the need for a tie-breaking mechanism. The temporal dimension of linked data is often challenging and some elements remain unresolved.

**Comment by steve-bate** [comment]
> The terms 'an object' and 'the object' in the AP specification don't necessarily imply functional constraints on the AS2 vocabulary.

Agreed. That's not what I was trying to communicate. ActivityPub is an extension of (and *apparently* a  restriction of) AS2. I am only referring to AP and the AP-specific usage of AS2, not AS2 *per se* which obviously could be used outside of an AP context.

**Comment by steve-bate** [comment]
> However, implementation practices may differ based on practical considerations, such as the cost of lookups and the need for a tie-breaking mechanism. The temporal dimension of linked data is often challenging and some elements remain unresolved.

I'm not questioning whether Mastodon should avoid a dereference if the object is embedded. I do think it should behave correctly when only an IRI is provided for a created object. It *already* does a dereference in that case but it doesn't use the dereferenced object to create the `Status` and instead tries to extract `Status` field data from the IRI string (which doesn't go well).

**Comment by bmottershead** [comment]
What text of the AP spec makes "object" functional though it isn't in AS?   If the AP spec writers intended to make AP a "constrained" implementation of AS, wouldn't they have spelled that out explicitly?

It seems more likely that it was an error or an oversight. If there was any "implicit" constraining of the AP spec going on, it resulted as a practical matter from how Mastodon and other developers were reading the specs.   And they read them in a way that made their implementations conforming.

**Comment by steve-bate** [comment]
> What text of the AP spec makes "object" functional though it isn't in AS? 

See my previous comment. If you're looking for an explicit statement in the spec, you're going to be disappointed.

> If the AP spec writers intended to make AP a "constrained" implementation of AS, wouldn't they have spelled that out explicitly?

One would hope, but apparently not. Feel free to ask the spec authors directly if you'd like. I'm certainly an unreliable source. In any case, this discussion is hijacking the original topic of this issue, so I won't comment further here.
